### PR TITLE
[WIP] Kernel refactoring

### PR DESCRIPTION
**Pull request by niboshi**
Refactor kernel code generation and invocation.

### Keys:
1. The code for launching kernels were duplicated (i.e. between `ElementwiseKernel` and `ufunc`, and between `ReductionKernel` and `simple_reduction_function`). I factored out common logic and merged the diverged code.
2. Encapsulated argument metadata (as `ArgInfo`). This is aimed at fusion code refactoring.
    * Divided `broadcast` to the metadata retrieval (`_broadcast_impl.__init__`) and application to actual arrays (`_broadcast_impl.apply`).
3. Refactored kernel code emission (with `KernelCodeEmitter`) for elementary operations (both elementwise and reduction) in modularized manner. This is also aimed at fusion.

### Current status:
Currently performance is degraded. That's why this PR is WIP.

Left column:  blocking time of Python code
Right column: GPU time measured by `cupy.cuda.get_elapsed_time`
(Measured with https://github.com/niboshi/cupy-perf)

```
Original [master @ 630ace19bfec4f61c9518401cbd8d577bdf5e765]
empty               :     0.200 us   +/- 0.014 us      5.343 us   +/- 0.410 us
sum                 :    35.549 us   +/- 0.987 us     42.337 us   +/- 1.176 us
sum_huge            :    42.373 us   +/- 9.522 us  10484.689 us   +/-14.838 us
add                 :    25.316 us   +/- 1.726 us     31.822 us   +/- 2.067 us
add_out             :    17.830 us   +/- 1.543 us     24.636 us   +/- 2.056 us
add_out_huge        :    21.851 us   +/- 4.426 us   9377.292 us   +/- 4.485 us
userkernel_create   :     2.353 us   +/- 0.625 us      7.757 us   +/- 1.014 us
userkenel           :    12.217 us   +/- 0.593 us     19.725 us   +/- 1.522 us
userkernel_huge     :    15.893 us   +/- 4.136 us   9371.349 us   +/- 4.243 us

```
```
This PR
empty               :     0.181 us   +/- 0.009 us      5.371 us   +/- 0.322 us
sum                 :    41.038 us   +/- 2.466 us     48.072 us   +/- 2.604 us
sum_huge            :    46.439 us   +/- 9.223 us  10484.955 us   +/-12.350 us
add                 :    29.894 us   +/- 0.631 us     36.599 us   +/- 0.706 us
add_out             :    21.656 us   +/- 0.450 us     28.692 us   +/- 0.596 us
add_out_huge        :    26.021 us   +/- 7.232 us   9381.202 us   +/- 7.555 us
userkernel_create   :     1.038 us   +/- 0.171 us      6.407 us   +/- 0.638 us
userkenel           :    15.110 us   +/- 1.232 us     22.115 us   +/- 1.913 us
userkernel_huge     :    18.648 us   +/- 5.055 us   9373.585 us   +/- 5.171 us
```
\* `empty` means empty Python function calls

### CODE REVIEW

**Comment by okuta** [comment]
IMO, this performance degradation can be accepted.

Please proceed with your work.

File: cupy/core/carray.pxi [hunk 1]
```diff
@@ -55,117 +45,23 @@ cdef class Indexer:
         self.shape = shape
         self.size = size
 
+    def __hash__(self):
+        return hash(self.size) ^ hash(self.shape)
```
Comment (okuta) [inline cupy/core/carray.pxi]: How about `return hash(self.shape)`?

File: cupy/core/carray.pxi [hunk 2]
```diff
@@ -55,117 +45,23 @@ cdef class Indexer:
...
+
+    def __richcmp__(Indexer x, Indexer y, int op):
+        if op == 2:
+            if x is y:
+                return True
+            return x.size == y.size and x.shape == y.shape
+
+        elif op == 3:
+            return not (x == y)
```
Comment (okuta) [inline cupy/core/carray.pxi]: Please write compere code directory.

File: cupy/core/compile.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+_header_path_cache = None
+_header_source = None
+
+
+def _get_header_dir_path():
+    global _header_path_cache
+    if _header_path_cache is None:
+        # Cython cannot use __file__ in global scope
+        _header_path_cache = os.path.abspath(
```
Comment (okuta) [inline cupy/core/compile.py]: We can use `__file__` in global scope on pure Python.

File: cupy/core/elementwise.pxi [hunk 4]
```diff
@@ -99,95 +74,69 @@ for i in six.integer_types:
...
-        dtype = numpy.dtype(dtype).type
-    return _typenames[dtype]
+    name = _typenames.get(dtype)
+    if name is None:
+        name = _typenames[numpy.dtype(dtype).type]
+    return name
 
 
-cpdef list _preprocess_args(args):
+cdef void _preprocess_args(list args, list arg_infos):
```
Comment (okuta) [inline cupy/core/elementwise.pxi]: Please remove `void`.

**Comment by okuta** [comment]
I want to merge this PR.
If you have no time, I can improve it.

**Comment by grafi-tt** [comment]
@niboshi I'd like to see a functionality to perform multiple reduction within one ReductionKernel. One notable use case is to compute L1 and L2 moments simultaneously. Currently, two invocation of ReductionKernel is required to compute variance and mean of data. (Example:  https://github.com/chainer/chainer/blob/v3.5.0/chainer/functions/normalization/layer_normalization.py#L39, where `mean` is called twice.) But it causes unnecessary memory accesses. Performance loss is not negligible, especially for ResNet-like CNNs where convolutions are relatively small.

Are you intended to implement this functionality in this PR?  If not, should I open an issue for the functionality?

Many thanks.

**Comment by niboshi** [comment]
Closing as improvements in this PR either has been already merged or being discussed in other PRs.
(e.g. #2801  #2920  #2940)

**Comment by niboshi** [comment]
@grafi-tt 
I'm sorry for a long absence in this PR...
This PR was only meant for refactoring, and no feature addition were intended.

**Pull request closed without merging**